src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -120,10 +120,26 @@ toTuple _ = error "Not a value declaration"
 makeCaseDeclaration :: (Functor m, Applicative m, MonadSupply m, MonadError MultipleErrors m) => Ident -> [([Binder], Either [(Guard, Expr)] Expr)] -> m Declaration
 makeCaseDeclaration ident alternatives = do
   let argPattern = length . fst . head $ alternatives
-  args <- map Ident <$> replicateM argPattern freshName
+  args <- mapM argName (foldl1 (zipWith resolveIdents) namedArgs)
   let
     vars = map (Var . Qualified Nothing) args
     binders = [ CaseAlternative bs result | (bs, result) <- alternatives ]
     value = foldr (Abs . Left) (Case vars binders) args
   return $ ValueDeclaration ident Value [] (Right value)
-
+  where
+  namedArgs = findNames . fst <$> alternatives
```

Comment:
Could you please add some type signatures here? You might need `ScopedTypeVariables`, I'm not sure.